airflow-core/src/airflow/cli/commands/variable_command.py

```
@@ -81,20 +81,25 @@ def variables_import(args, session):
     """Import variables from a given file."""
     if not os.path.exists(args.file):
         raise SystemExit("Missing variables file.")
-    with open(args.file) as varfile:
-        try:
-            var_json = json.load(varfile)
-        except JSONDecodeError:
-            raise SystemExit("Invalid variables file.")
+
+    try:
+        # Use load_variables which supports JSON, YAML, and ENV formats
+        var_dict = load_variables(args.file)
+    except Exception as e:
```

Is it possible to catch a narrower exception?

---

I think this is not resolved. What I meant is whether we could change `except Exception` to something like `except JSONDecodeError` but contains all the possible exceptions expected

---

I guess that one still needs to be resolved?

---

yep, I think this is still missed

---

I've implemented the requested alignment between airflow variables import and `LocalFilesystemBackend` using the centralized `load_variables()` function. However, during development I discovered an important limitation:
**Breaking Change Identified:**
The current load_variables() implementation in LocalFilesystemBackend doesn't support the description property for variables. This means that moving the CLI import to use this shared logic removes the existing ability to import variables with descriptions from **JSON files**.

**Current State:**
  - ✅ CLI now uses centralized error handling and supports YAML/ENV formats
  - ✅ Better error messages and file format validation
  - ❌ Lost capability: JSON files with variable descriptions no longer work
  - ❌ No test coverage: I couldn't find tests validating description imports

  **Decision Needed:**
  We need to decide the priority between:
  1. New capability: YAML and ENV file support for variable imports
  2. Existing capability: JSON files with variable descriptions

  **Options:**
  - Accept the breaking change and document it
  - Enhance LocalFilesystemBackend.load_variables() to support descriptions
  - Revert to custom JSON parsing in CLI while keeping error handling improvements

Please advise on the preferred approach. The current implementation is ready but introduces this breaking change.

P.S. @Lee-W, parsing level exceptions `yaml.MarkedYAMLError`, `JSONDecodeError` handled in parser functions `_parse_yaml_file` and `_parse_json_file` respectively and transformed into `AirflowFileParseException` later. As for FileNotFound and other OSErrors they are handled directly in `variables_import` function.

---

I also found this old https://github.com/apache/airflow/issues/37532 which shows that working with variable descriptions had limitations even before this change. The issue highlighted that variable exports from the UI didn't include description fields, requiring manual updates when importing to other environments. This suggests the description functionality was already incomplete in the broader ecosystem.

Happy to tackle the description functionality improvements if you're interested and can point me in the right direction.

---

>    Enhance LocalFilesystemBackend.load_variables() to support descriptions
 

ideally this

---

> parsing level exceptions yaml.MarkedYAMLError, JSONDecodeError handled in parser functions _parse_yaml_file and _parse_json_file respectively and transformed into AirflowFileParseException later. As for FileNotFound and other OSErrors they are handled directly in variables_import function.

Yep, but fail early is still better

---

> Happy to tackle the description functionality improvements if you're interested and can point me in the right direction.

Not that familiar with this piece, but I would try to see how I can help when I have more bandwidth. Thanks!

---

> Yep, but fail early is still better

I'm not sure I've got you point. Could you please clarify what exactly do you mean?

Current error handling strategy (as I understand it) is to catch parsing errors from 3 low-level parser functions `_parse_<env|json|yaml>_file`, repackage and return them as lists of `FileSyntaxError` to the caller `_parse_secret_file` and let the caller raise the `AirflowFileParseException` with all collected errors, which let the users see all the issues in their configuration files at once rather than making them fix one by one.

---

Got it. I just took a deeper look. I just noticed all the cases you mentioned are handled above. Do we still need this `except Exception as e:` block?

---

As we work with user files, I think it's still useful for catching these exceptions: 
- PermissionError (file access issues)
- OSError (other I/O errors beyond FileNotFoundError)
- MemoryError (for very large files)
- UnicodeDecodeError (encoding issues)

I can see that pattern in connection_command too.

---

I still like catching explicit exceptions more, but if this pattern is used elsewhere. I'm also ok with it

---

I think it's pretty common pattern in CLI tools. I personally do not like to have ugly (nor even nice) stack trace as result of unhandled exception in CLI so try/except Exception allows to have a nice one line exception message - ideally in 🟥 ....